src/ui/public/share/directives/share.js

```diff
@@ -1,18 +1,149 @@
+import { parse as parseUrl } from 'url';
+
+import {
+  getUnhashableStatesProvider,
+  unhashUrl,
+} from 'ui/state_management/state_hashing';
+import Notifier from 'ui/notify/notifier';
+
+import urlShortenerProvider from '../lib/url_shortener';
+
 import uiModules from 'ui/modules';
 import shareTemplate from 'ui/share/views/share.html';
 const app = uiModules.get('kibana');
 
-app.directive('share', function () {
+app.directive('share', function (Private) {
+  const getUnhashableStates = Private(getUnhashableStatesProvider);
+  const urlShortener = Private(urlShortenerProvider);
+
   return {
     restrict: 'E',
     scope: {
       objectType: '@',
       objectId: '@',
-      setAllowEmbed: '&?allowEmbed'
+      allowEmbed: '@',
     },
     template: shareTemplate,
-    controller: function ($scope) {
-      $scope.allowEmbed = $scope.setAllowEmbed ? $scope.setAllowEmbed() : true;
+    controllerAs: 'share',
+    controller: function ($scope, $document, $location, globalState) {
+      if ($scope.allowEmbed !== 'false' && $scope.allowEmbed !== undefined) {
+        throw new Error('allowEmbed must be "false" or undefined');
+      }
+
+      // Default to allowing an embedded IFRAME, unless it's explicitly set to false.
+      this.allowEmbed = $scope.allowEmbed === 'false' ? false : true;
+      this.objectType = $scope.objectType;
+
+      function getOriginalUrl() {
+        // If there is no objectId, then it isn't saved, so it has no original URL.
+        if ($scope.objectId === undefined || $scope.objectId === '') {
+          return;
+        }
+
+        // Remove the query string.
+        const baseUrl = $location.absUrl().split('?')[0];
+
+        // Add global state to the URL so that the iframe doesn't just show the time range default.
+        return `${baseUrl}?_g=${$location.search()._g}`;
```

Comment:
Oh, I see what you're saying! I forgot about that gotcha about the way Angular stores stuff after the hash. Great call, thanks.